design/design.md

```diff
@@ -0,0 +1,380 @@
+
+The coroutines posted to the `JobQueue` populate the response and submit the
+response for sending. Once gRPC actually sends
+the response (gRPC handles the sending internally), an event is placed on the `CompletionQueue`; when
+`CompletionQueue::Next()` returns such an event, signaling the response has been
+sent, the event loop cleans up the resources associated with this
+request/response.
+
+While events are consumed from the `CompletionQueue` sequentially, responses are
+populated and sent in parallel. This is because responses are populated and sent
+from coroutines executed by the worker threads of the `JobQueue`, of which there are
+several (meaning several coroutines could be executing in
+parallel), and the event loop does not wait for the coroutines to execute in
+order to post more coroutines corresponding to additional requests.
+`CompletionQueue` is thread safe.
+
+The `ResourceManager` of rippled will be used to track resource consumption per
+endpoint. If the `ResourceManager` indicates that we should "disconnect" from
+a specific endpoint, a coroutine will not be posted to the `JobQueue`, the
+handler will not be called and an error will be returned to the client.
+
+## Classes
+The server code is adapted from the example found
+[here](https://github.com/grpc/grpc/blob/v1.24.0/examples/cpp/helloworld/greeter_async_server.cc)
+
+`CompletionQueue` is a class of the GRPC library. `CompletionQueue` is used to
+queue events, and provides a method, `CompletionQueue::Next()` to query for any
+events that have occurred. Events are returned by `CompletionQueue` in the form
+of a `void **` pointer, known as a tag. In our application, `CompletionQueue` returns pointers to
+`CallData` objects, which are used to service requests.
+
+Each request is represented by a `CallData` object. `CallData` is an abstract
+class that implements a method `Proceed()` and can be in one of three states:
+LISTEN, PROCESS or FINISH. These states correspond to different states in the
+lifecycle of an rpc call, and `Proceed()` transitions the object from one state
+to the next. `CallData` has two pure virtual methods: `makeListener()` and
+`process()`, which are implemented by derived classes. For
+each rpc method, there is a class that derives from `CallData`, and implements
+the pure virtual methods. `finish()` can be overridden if necessary. The
+destructor is virtual so that `delete this` actually deletes the entire object,
+including the derived class.
+
+**In the below example, some code is ommitted for the sake of clarity and
+brevity, including arguments to functions and member variables**
+
+```
+class CallData {
+
+    CallData(CompletionQueue* cq) :cq_(cq), _status(LISTEN)
+    {
+    }
+
+    virtual ~CallData() {}
+    void Proceed() {
+      if (status_ == LISTEN) {
+          status_ = PROCESS;
+          process();
+      } else {
+          _status = FINISH;
+          finish();
+          delete this;
+      }
+    }
```

Comment:
I'd remove `Proceed` and have the `EvenLoop` call `process` if needed.

Reply:
changed so event loop calls process